src/main/cpp/src/timezones.cu

```diff
@@ -118,6 +119,221 @@ auto convert_timestamp_tz(column_view const& input,
+    bool const is_gap       = (upper_bound_utc + to_utc_offset == upper_bound_epoch);
+    if (!is_gap && upper_bound_utc != std::numeric_limits<int64_t>::min() &&
+        upper_bound_utc != std::numeric_limits<int64_t>::max()) {  // overlap
+      // The overlap range is [utcInstant + offsetBefore, utcInstant + offsetAfter]
+      auto const overlap_before = static_cast<int64_t>(upper_bound_utc + to_utc_offset);
+      if (result_epoch_seconds >= overlap_before && result_epoch_seconds <= upper_bound_epoch) {
+        // By default, to_utc_offset is the offsetAfter, so if the to_local_offset is valid and
+        // need to replace the to_utc_offset, it only happens when it is early_offset
+        if (early_offset == prefered_offset) { to_utc_offset = early_offset; }
+      }
+    }
+
+    auto const to_utc_offset_duration = cuda::std::chrono::duration_cast<duration_type>(
+      cudf::duration_s{static_cast<int64_t>(to_utc_offset)});
+
+    // subtract the offset to convert local to utc
+    return timestamp - to_utc_offset_duration;
+  }
+
+  __device__ inline cudf::timestamp_us plus_with_tz(cudf::timestamp_us const& timestamp,
+                                                    cudf::duration_us const& duration) const
+  {
+    if (duration == cudf::duration_us{0}) { return timestamp; }
+
+    auto const utc_instants = transitions.child().child(0);
+    auto const tz_instants  = transitions.child().child(1);
+    auto const utc_offsets  = transitions.child().child(2);
+
+    auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+    auto const list_size      = tz_transitions.size();
+
+    auto const transition_times_utc = cudf::device_span<int64_t const>(
+      utc_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const transition_times_tz = cudf::device_span<int64_t const>(
+      tz_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const transition_offsets = cudf::device_span<int32_t const>(
+      utc_offsets.data<int32_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    // In Spark, timeAdd will add the days of the duration to the timestamp first with `plusDays`,
+    // resolve the offset if the result is in the overlap, and then add the remainder microseconds
+    // to the results with `plus`, then resolve again. In java, the `plusDays` adds the days to the
+    // localDateTime in day's field, so we need take transitions into account. While the `plus` step
+    // adds the remainder microseconds to the localDateTime in microsecond's field, so it is
+    // equivalent to add the remainder to the utc result without resolving the offset.
+    auto const duration_value =
+      static_cast<int64_t>(cuda::std::chrono::duration_cast<cudf::duration_us>(duration).count());
+    auto const microseconds_per_day = 86400000000ll;
```

Comment:
Have we tested with more microseconds than in a day, along with a days value? I know that this is an odd corner case, but if we have to apply days separately from microseconds wouldn't this result in the same kind of errors if we assume that we can always `convert microseconds_per_day` into days? It looks like for DayTimeIntervalType microsecond values outside of a day are not technically allowed, so we are probably okay there, but CalendarIntervalType does not indicate any limitations like that, and reading the code there are none.

We might be able to work around this by restricting the usage of CalendarInterval, but I'd prefer not to need to do that.

Reply:
Following test can pass with this pr:
```
@pytest.mark.parametrize('data_gen', [(0, 86400000001), (1, 86400000001), (1, 86400000000*100)], ids=idfn)
@allow_non_gpu(*non_supported_tz_allow)
def test_timeadd_debug(data_gen):
    days, microseconds = data_gen
    assert_gpu_and_cpu_are_equal_collect(
        lambda spark: unary_op_df(spark, TimestampGen(), length=200000)
            .selectExpr("a + (interval {} days {} microseconds)".format(days, microseconds)))
```

It's an interesting case I hadn't thought of, I'll check the code to see if there is some validation in spark/pyspark to pass it. 

If it is possible to pass an invalid CalendarInterval, we can have a simple workaround in plugin side to pass only the days in this kernel and add the microseconds directly to the result.

Reply:
Ok, this case failed on 311. Spark seems to automatically use DayTimeIntervalType for interval literals in higher versions. Will match this behavior in plugin PR.

Reply:
Matched this in plugin datetimeExpressionsUtils.scala